distribution/docs/src/main/jdocs/content/_developingComponents/developing-plugins-contents.adoc

```diff
@@ -0,0 +1,403 @@
+
+|<<_post_resource_plugins,Post-Resource>>
+|ddf.catalog.plugin.PostResourcePlugin
+|After a Resource is retrieved, but before it is sent to the Endpoint.
+
+|<<_pre_create_storage_plugins,Pre-Create Storage>>
+|ddf.catalog.plugin.PreCreatePlugin
+|Before an item is created in the content repository.
+
+|<<_post_create_storage_plugins,Post-Create Storage>>
+|ddf.catalog.plugin.PostResourcePlugin
+|After an item is created in the content repository.
+
+|<<_pre_update_storage_plugins,Pre-Update Storage>>>
+|ddf.catalog.plugin.PostResourcePlugin
+|Before an item is updated in the content repository.
+
+|<<_post_update_storage_plugins,Post-Update Storage>>>
+|ddf.catalog.plugin.PostResourcePlugin
+|After an item is updated in the content repository.
+
+|<<_pre_subscription_plugins,Pre-Subscription>>
+|ddf.catalog.plugin.PreSubscriptionPlugin
+|Prior to a Subscription being created or updated.
+
+|<<_pre_delivery_plugins,Pre-Delivery>>
+|ddf.catalog.plugin.PreDeliveryPlugin
+|Prior to the delivery of a Metacard when an event is posted.
+
+|===
+
+==== Implement Plugins
+
+The procedure for implementing any of the plugins follows a similar format:
+
+. Create a new class that implements the specified plugin interface.
+
+. Implement the required methods.
+
+. Create an OSGi descriptor file to communicate with the OSGi registry.
+
+.. Register the plugin class as a service to OSGi registry.
+
+. Deploy to ${branding}.
+
+.Plugin Performance Concerns
+[NOTE]
+====
+Plugins should include a check to determine if requests are local or not. It is usually preferable to take no action on non-local requests.
+====
+
+[TIP]
+====
+Refer to the Javadoc for more information on all Requests and Responses in the `${ddf-branding-lowercase}.catalog.operation` and `${ddf-branding-lowercase}.catalog.event` packages.
+====
+
+===== Implementing Pre-Ingest Plugins
+
+Develop a custom Pre-Ingest Plugin.
+
+. Create a Java class that implements `PreIngestPlugin`. +
+`public class SamplePreIngestPlugin implements ddf.catalog.plugin.PreIngestPlugin`
++
+. Implement the required methods. +
+ * `public CreateRequest process(CreateRequest input) throws PluginExecutionException, StopProcessingException;`
+ * `public UpdateRequest process(UpdateRequest input) throws PluginExecutionException, StopProcessingException;`
+ * `public DeleteRequest process(DeleteRequest input) throws PluginExecutionException, StopProcessingException;`
++
+. Import the ${branding} interface packages to the bundle manifest (in addition to any other required packages). +
+`Import-Package: ddf.catalog,ddf.catalog.plugin`
++
+. Export the service to the OSGi registry. +
+*Blueprint descriptor example*
+`<service ref="[[SamplePreIngestPlugin ]]"interface="ddf.catalog.plugin.PreIngestPlugin" />`
+
+====== Pre-Ingest Plugin Failure Behavior
+
+In the event that this Catalog Plugin cannot operate but does not wish to fail the transaction, a `PluginExecutionException` will be thrown.
```

Comment:
Also, would it be more logical to mention the two exceptions plug-ins can throw first (`PlugInExecutionException` and `StopProcessingException`) and then say what happens for any other exceptions?
Same for all those sections below.